### PR TITLE
Make our VMs internal, hiding their applications from the menu

### PR DESCRIPTION
**Pull request by eaon**
## Description of Changes

We want our users to interact with the system through the client, which is started via the updater, rather than directly with a VM which would be the normal Qubes way. This should make the menu a bit less noisy as well.

The omission of `sd-whonix` is intentional, as it contains a couple of Tor related GUI tools that users might need access to and have no other way to get to.

**Todo:** add tests.

## Testing

- [ ] After running `sdw-admin --apply` / `make dev`, all of our VMs with the exception of `sd-whonix` and `sd-devices` are hidden from the Qubes menu.
- [ ] Tests in `dom0` pass

### CODE REVIEW

**Comment by cfm** [comment]
Just flagging here, @eaon, the following caveat on this approach in QubesOS/qubes-issues#8042:

> [...] I would think that trying to use any future "hide internal qubes in Qube Manager" feature in order to hide _other_ types of qubes would likely constitute an unintended and out-of-scope (ab)use of that feature. [...]
> 
> (This is based on the assumption that "internal" qubes are specified by the system itself and aren't intended to be user-specifiable. If my assumption is wrong, and the Qubes developers _do_ intend for administrators or users to be able to create their own "internal" qubes, then perhaps this is the right issue after all.)

**Comment by cfm** [comment]
The Qubes team confirmed yesterday that `internal` is not reserved for "Qubes OS official use only", so we're free to use this flag to hide our VMs from the Qubes menu and the Qube Manager.  (They'll be visible only in the running "Qubes Domains" menu".)

**Review by nathandyer — CHANGES_REQUESTED** [review]
I can confirm that this is mostly working as expected on a Qubes 4.1 system with SecureDrop Workstation installed, after setting up a dev environment, checking out the branch, and running `make clone` and `make dev` from `dom0`. sd-app, sd-gpg, sd-log, sd-proxy, and sd-viewer are all hidden from the menu, as expected. Likewise, sd-whonix is still showing up as intended.

For reasons that aren't yet clear to me, sd-devices is still showing up in the menu, even though it should be hidden, and is set using the same mechanism as the others. Perhaps that's a quirk of my system, but I don't think we can call this good until that is sorted out.

The code LGTM. We should be OK to merge once the sd-devices menu item is figured out.

**Comment by rocodes** [comment]
Reviving this to discuss and either include or close. Here's the relevant 4.2 behaviour, per marta:
 
> *    internal qubes will be filtered out by default in qube manager, but can be shown in the 'View' menu
> *    they will show up in the qui-domains widget, but the menu will be truncated (just log + shutdown/kill) and have an explicit "Internal qube" label with a tooltip explanation to the tune of "Please don't mess with it if you aren't 100% sure you know what you're doing".
> 
https://github.com/QubesOS/qubes-issues/issues/8064#issuecomment-1478455721

This means:
* No opening a terminal, File Manager, or Qubes Settings etc via top righthand Qubes (qui-domains) widget -> qube  (That would have to be done eg via `qvm-run` from dom0 terminal)
* You _can_ grab a console log and copy it to the global clipboard on internal qubes with one click from that menu (cool)  
* You can shutdown an internal qube from the menu (shutdown, not kill). To restart you'd have to use dom0 commands.
* This has nothing to do with Devices widget and which qubes are displayed as attach targets 

tldr: 
This is still a nice way to clean up the menu for qubes we don't want users interacting with directly... as long as none of our troubleshooting steps ask them to shutdown or restart the VM via the GUI menus.

**Comment by zenmonkeykstop** [comment]
It also means we can maybe drop a bunch of appmenu configuration, as we would not be using them in normal operations.

**Comment by rocodes** [comment]
Ok. So IMO, sd-whonix and sd-devices should not be included (too many times we may need to interact with them), but anything else is fair game. I'll update

**Review by legoktm — DISMISSED** [review]
on Qubes 4.2 the VMs initially show up in the menu (despite e.g. `qvm-features sd-app internal` returning 1) and are only hidden after a reboot. I think that's reasonable for our purposes since we mandate a reboot after install. Tests pass too.

My only ask/question is if whether we can leave them in the VM list for dev installs. I'm used to opening terminals, etc. in VMs and debugging things and would be nice if we didn't need to launch them through qvm-run, etc.

**Comment by rocodes** [comment]
Yeah sure, that makes sense to me. Will update later today, thanks for review

**Comment by rocodes** [comment]
```
Configure base template
INFO:[2024-05-08-15:34:44:685309] [ERROR   ] Command '/tmp/__salt.tmp.poqwzl3a' failed with return code: 1
INFO:[2024-05-08-15:34:44:685340] [ERROR   ] stdout: update-xfce-settings: Adjusting icon size for user user to 64 px
INFO:[2024-05-08-15:34:44:685368] [ERROR   ] stderr: Failed to init libxfconf: Could not connect: No such file or directory.
INFO:[2024-05-08-15:34:44:685399] [ERROR   ] retcode: 1
INFO:[2024-05-08-15:34:44:685426] [ERROR   ] {'pid': 6928, 'retcode': 1, 'stdout': 'update-xfce-settings: Adjusting icon size for user user to 64 px', 'stderr': 'Failed to init libxfconf: Could not connect: No such file or directory.'}
```
Slightly strange ci behaviour here

**Comment by legoktm** [comment]
> Slightly strange ci behaviour here

This is #1006 - no ideas yet.

**Comment by legoktm** [comment]
```
INFO:[2024-05-14-22:11:24:163934]           ID: sd-proxy
INFO:[2024-05-14-22:11:24:163968]     Function: qvm.vm
INFO:[2024-05-14-22:11:24:164001]       Result: False
INFO:[2024-05-14-22:11:24:164035]      Comment: An exception occurred in this state: Traceback (most recent call last):
INFO:[2024-05-14-22:11:24:164064]                 File "/usr/lib/python3.11/site-packages/salt/state.py", line 2423, in call
INFO:[2024-05-14-22:11:24:164100]                   ret = self.states[cdata["full"]](
INFO:[2024-05-14-22:11:24:164133]                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164166]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 159, in __call__
INFO:[2024-05-14-22:11:24:164199]                   ret = self.loader.run(run_func, *args, **kwargs)
INFO:[2024-05-14-22:11:24:164230]                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164265]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1245, in run
INFO:[2024-05-14-22:11:24:164299]                   return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
INFO:[2024-05-14-22:11:24:164333]                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164366]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1260, in _run_as
INFO:[2024-05-14-22:11:24:164403]                   return _func_or_method(*args, **kwargs)
INFO:[2024-05-14-22:11:24:164436]                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164470]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1293, in wrapper
INFO:[2024-05-14-22:11:24:164505]                   return f(*args, **kwargs)
INFO:[2024-05-14-22:11:24:164546]                          ^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164582]                 File "/var/cache/salt/minion/extmods/states/ext_state_qvm.py", line 431, in vm
INFO:[2024-05-14-22:11:24:164613]                   _varargs, keywords = parse_options(kwargs[action])
INFO:[2024-05-14-22:11:24:164647]                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-05-14-22:11:24:164676]                 File "/var/cache/salt/minion/extmods/states/ext_state_qvm.py", line 419, in parse_options
INFO:[2024-05-14-22:11:24:164711]                   for option in options:
INFO:[2024-05-14-22:11:24:164740]               TypeError: 'NoneType' object is not iterable
```

File: securedrop_salt/sd-proxy.sls [hunk 1]
```diff
@@ -23,13 +24,18 @@ sd-proxy-dvm:
...
+    {% endif %}
     - tags:
       - add:
         - sd-workstation
         - sd-{{ sdvars.distribution }}
     - require:
-      - qvm: sd-whonix
-      - qvm: sd-small-{{ sdvars.distribution }}-template
+      - qvm: securedrop_salt.sd-whonix
+      - qvm: securedrop_salt.sd-small-{{ sdvars.distribution }}-template
```
Comment (cfm) [inline securedrop_salt/sd-proxy.sls:38]: Based on—

https://github.com/freedomofpress/securedrop-workstation/blob/c13aeaa500dc3cbce58c45e20dbc1c5e9d07b55e/securedrop_salt/sd-proxy.sls#L13-L15

—I believe this needs to be:

```suggestion
      - sls: securedrop_salt.sd-whonix
      - sls: securedrop_salt.sd-small-{{ sdvars.distribution }}-template
```

<details><summary>Failure on previous commit</summary></details>

```
INFO:[2024-06-04-18:56:49:482843]           ID: sd-proxy-dvm
INFO:[2024-06-04-18:56:49:482867]     Function: qvm.vm
INFO:[2024-06-04-18:56:49:482894]       Result: False
INFO:[2024-06-04-18:56:49:482919]      Comment: An exception occurred in this state: Traceback (most recent call last):
INFO:[2024-06-04-18:56:49:482947]                 File "/usr/lib/python3.11/site-packages/salt/state.py", line 2423, in call
INFO:[2024-06-04-18:56:49:482971]                   ret = self.states[cdata["full"]](
INFO:[2024-06-04-18:56:49:483001]                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483025]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 159, in __call__
INFO:[2024-06-04-18:56:49:483051]                   ret = self.loader.run(run_func, *args, **kwargs)
INFO:[2024-06-04-18:56:49:483078]                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483102]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1245, in run
INFO:[2024-06-04-18:56:49:483127]                   return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
INFO:[2024-06-04-18:56:49:483154]                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483179]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1260, in _run_as
INFO:[2024-06-04-18:56:49:483206]                   return _func_or_method(*args, **kwargs)
INFO:[2024-06-04-18:56:49:483240]                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483269]                 File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1293, in wrapper
INFO:[2024-06-04-18:56:49:483297]                   return f(*args, **kwargs)
INFO:[2024-06-04-18:56:49:483321]                          ^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483346]                 File "/var/cache/salt/minion/extmods/states/ext_state_qvm.py", line 431, in vm
INFO:[2024-06-04-18:56:49:483372]                   _varargs, keywords = parse_options(kwargs[action])
INFO:[2024-06-04-18:56:49:483397]                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-18:56:49:483421]                 File "/var/cache/salt/minion/extmods/states/ext_state_qvm.py", line 419, in parse_options
INFO:[2024-06-04-18:56:49:483449]                   for option in options:
INFO:[2024-06-04-18:56:49:483473]               TypeError: 'NoneType' object is not iterable
INFO:[2024-06-04-18:56:49:483501]      Started: 18:56:26.557304
INFO:[2024-06-04-18:56:49:483529]     Duration: 2409.68 ms
INFO:[2024-06-04-18:56:49:483555]      Changes:   
```

</details>

<details><summary>Failure on this commit</summary>

```
INFO:[2024-06-04-21:23:59:145074]           ID: sd-proxy-dvm
INFO:[2024-06-04-21:23:59:145099]     Function: qvm.vm
INFO:[2024-06-04-21:23:59:145126]       Result: False
INFO:[2024-06-04-21:23:59:145154]      Comment: The following requisites were not found:
INFO:[2024-06-04-21:23:59:145180]                                  require:
INFO:[2024-06-04-21:23:59:145206]                                      qvm: securedrop_salt.sd-whonix
INFO:[2024-06-04-21:23:59:145234]                                      qvm: securedrop_salt.sd-small-bookworm-template
INFO:[2024-06-04-21:23:59:145386]      Started: 21:23:37.906327
INFO:[2024-06-04-21:23:59:145417]     Duration: 0.003 ms
INFO:[2024-06-04-21:23:59:145443]      Changes:
```

</details>

**Review by cfm — CHANGES_REQUESTED** [review]

File: securedrop_salt/sd-proxy.sls (hunk 1 above)
Comment (rocodes) [reply securedrop_salt/sd-proxy.sls:38]: ugh. thank you @cfm !

Comment (rocodes) [reply securedrop_salt/sd-proxy.sls]: (update: actually see [below](https://github.com/freedomofpress/securedrop-workstation/pull/857#issuecomment-2149657488): in this case we do want to rely on just the vm creation as a prereq (qvm), not the entire sls file )

**Comment by zenmonkeykstop** [comment]
Just realized something - I don't think we can verify this properly until there are prod bookworm packages on apt.freedom.press, as the menu changes are being applied per-env, and we can't do a full prod deployment yet :/. Setting the internal feature at the dom0 commandline works tho.

**Review by zenmonkeykstop — APPROVED** [review]
- [x] `make dev`  passing
- [x] CI and visual review are good
